src/consensus/pool/finality_tracker.rs

```diff
@@ -0,0 +1,344 @@
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_fast_finalized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::FinalizedAndNotarized(_) => {
+                    return FinalizationEvent::default();
+                }
+                FinalizationStatus::Notarized(_)
+                | FinalizationStatus::Finalized
+                | FinalizationStatus::ImplicitlyFinalized(_) => {}
+                FinalizationStatus::ImplicitlySkipped => unreachable!("consensus safety violation"),
+            }
+        }
+
+        let mut event = FinalizationEvent::default();
+        self.status
+            .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+        self.handle_finalized((slot, block_hash), &mut event);
+        self.highest_finalized = slot.max(self.highest_finalized);
+        event
+    }
+
+    /// Marks the given block as notarized.
+    ///
+    /// Handles possibly resulting direct finalization of the block.
+    /// Further, also handles any possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_notarized(&mut self, slot: Slot, block_hash: Hash) -> FinalizationEvent {
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::Notarized(_)
+                | FinalizationStatus::FinalizedAndNotarized(_)
+                | FinalizationStatus::ImplicitlyFinalized(_)
+                | FinalizationStatus::ImplicitlySkipped => FinalizationEvent::default(),
+                FinalizationStatus::Finalized => {
+                    let mut event = FinalizationEvent::default();
+                    self.status
+                        .insert(slot, FinalizationStatus::FinalizedAndNotarized(block_hash));
+                    self.handle_finalized((slot, block_hash), &mut event);
+                    event
+                }
+            }
+        } else {
+            self.status
+                .insert(slot, FinalizationStatus::Notarized(block_hash));
+            FinalizationEvent::default()
+        }
+    }
+
+    /// Marks the given slot as finalized.
+    ///
+    /// Handles possibly resulting direct finalization of a block in this slot.
+    /// Further, also handles any possibly resulting implicit finalizations.
+    ///
+    /// Returns a [`FinalizationEvent`] that contains information about newly finalized slots.
+    pub fn mark_finalized(&mut self, slot: Slot) -> FinalizationEvent {
+        if let Some(status) = self.status.get(&slot) {
+            match status {
+                FinalizationStatus::Finalized
+                | FinalizationStatus::FinalizedAndNotarized(_)
```

Comment:
We can get multiple finalizations for a slot if we got fast finalization and slow one?  Are there other scenarios?

Reply:
yes, if `FinalityTracker` is used correctly there should be at most two triggers to finalize a block in a given slot, however there are potentially many ways finalization can be triggered within pool (constructing FastFinal cert ourselves, receiving it, constructing Notar+Final cert ourselves, receiving those, and then there are many different certs) I think low-level components like `FinalityTracker` (and other parts of the code) ideally should not break if these events are triggered more than once by some higher-level component